src/Elm/Parser/Expression.elm

```diff
@@ -1,45 +1,220 @@
 import Elm.Syntax.Range exposing (Location, Range)
 import Elm.Syntax.Signature exposing (Signature)
-import Parser as Core exposing (Nestable(..))
-
-
-expressionNotApplication : Parser State (Node Expression)
-expressionNotApplication =
-    lazy
-        (\() ->
-            oneOf
-                [ numberExpression
-                , referenceExpression
-                , ifBlockExpression
-                , tupledExpression
-                , recordAccessFunctionExpression
-                , operatorExpression
-                , letExpression
-                , lambdaExpression
-                , literalExpression
-                , charLiteralExpression
-                , recordExpression
-                , glslExpression
-                , listExpression
-                , caseExpression
-                ]
-                |> Combine.andThen liftRecordAccess
+import Parser as Core exposing ((|=), Nestable(..))
+import Pratt exposing (Config)
+
+
+
+--expressionNotApplication : Parser State (Node Expression)
+--expressionNotApplication =
+--    lazy
+--        (\() ->
+--            oneOf
+--                [ numberExpression
+--                , referenceExpression
+--                , ifBlockExpression
+--                , tupledExpression
+--                , recordAccessFunctionExpression
+--                , operatorExpression
+--                , letExpression
+--                , lambdaExpression
+--                , literalExpression
+--                , charLiteralExpression
+--                , recordExpression
+--                , glslExpression
+--                , listExpression
+--                , caseExpression
+--                ]
+--                |> Combine.andThen liftRecordAccess
+--        )
+
+
+expression : Parser State (Node Expression)
+expression =
+    Pratt.expression
+        { oneOf =
+            [ numberExpression
+                |> Pratt.literal
+            , referenceExpression
+                |> Pratt.literal
+            , ifBlockExpression
+            , tupledExpression
+            , recordAccessFunctionExpression
+                |> Pratt.literal
+            , negationOperation
+            , letExpression
+            , lambdaExpression
+            , literalExpression
+                |> Pratt.literal
+            , charLiteralExpression
+                |> Pratt.literal
+            , recordExpression
+            , glslExpression
+                |> Pratt.literal
+            , listExpression
+            , caseExpression
+            ]
+        , andThenOneOf =
+            [ -- Elm Basics sets the precedence of <| and |> to 0, but Pratt.expression starts at 0 so it needs to be larger
+              infixRight 1 "<|"
+            , infixLeft 1 "|>"
+            , infixRight 2 "||"
+            , infixRight 3 "&&"
+
+            -- TODO Report a syntax error when encountering multiple of the comparison operators
+            -- `a < b < c` is not valid Elm syntax
```

Comment:
I'm on the fence about this. `a < b < c` seems like something that should parse correctly but be rejected at the next step by the compiler, a little like if you tried to define two functions with the same name.
I could be wrong though, it's right at the line between syntax and AST verification, and the Elm compiler doesn't really separate the steps (I haven't looked at the source).

Reply:
You're right that (as far as I know) it's done at a further step.

But I would be in favor of supporting it (if it's not too perf-draining). The point is that this syntax will always be incorrect (regardless of what kind of error the compiler reports), because it would be useful for tools using this library to know when they look (elm-review, ...) at or generate (elm-codegeni think, elm-review, ...) incorrect code. Right now you could create an autofix in elm-review which would be incorrect, whereas syntax errors are prevented by elm-review.

Similarly, I think we could check when `|>` and `<|` are used together like `a |> b <| c`.

But I'm happy for this to be a stretch goal, done even after releasing the pratt-pqrser change. And if it impacts performance or complexity too much, then I'm considering not doing it.